### PR TITLE
RuleEngine & RuleRegistry became RegistryChangeListener, ResourceBund…

### PR DESCRIPTION
**Pull request by adimova**
…le Providers implement Provider<E> #2361

Signed-off-by: Ana Dimova <a.dimova@prosyst.bg>

### CODE REVIEW

**Comment by adimova** [comment]
This PR is a second step of #2361 implementation. ResourceBundleProviders are with temporary variant of the code, until the ModuleTypeProvider and TemplateProvider interfaces are changed to extend the Provider interface.

File: bundles/automation/org.eclipse.smarthome.automation.core.test/src/test/java/org/eclipse/smarthome/automation/core/internal/RuleEngineTest.java [hunk 1]
```diff
@@ -54,19 +50,21 @@ private RuleEngine createRuleEngine() {
     @Test
     public void testAddRetrieveRules() {
         RuleEngine ruleEngine = createRuleEngine();
-        Rule rule0 = new Rule();
+        Rule rule0 = ruleEngine.initRuleId(new Rule());
         ruleEngine.addRule(rule0, true);
-        Collection<Rule> rules = ruleEngine.getRules();
+        Collection<RuntimeRule> rules = ruleEngine.rules.values();
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core.test/src/test/java/org/eclipse/smarthome/automation/core/internal/RuleEngineTest.java]: Don't use rules.values. It should be private field. Do a method getRuntimeRules() which has to return unmodifiable collection of rules

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 2]
```diff
@@ -118,42 +128,24 @@
...
-
-    /**
      * {@link Map} system module type to corresponding module handler factories.
      */
     private Map<String, ModuleHandlerFactory> moduleHandlerFactories;
 
     /**
-     * Locker which does not permit rule initialization when the rule engine is stopping.
-     */
-    private boolean isDisposed = false;
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: Don't remove this flag. It signals that the dispose operation is started and no more rule initializations are accepted

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 3]
```diff
@@ -167,23 +159,96 @@ public RuleEngine() {
         this.moduleHandlerFactories = new HashMap<String, ModuleHandlerFactory>(20);
     }
 
-    protected void setModuleTypeRegistry(ModuleTypeRegistry moduleTypeRegistry) {
-        this.mtRegistry = moduleTypeRegistry;
-        ConnectionValidator.setRegistry(mtRegistry);
+    @Override
+    public void added(ModuleType moduleType) {
+        Set<String> notInitailizedRules = new HashSet<String>(20);
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: The notInitailizedRules  set is not needed.

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 4]
```diff
@@ -167,23 +159,96 @@ public RuleEngine() {
...
+            for (String rUID : rules) {
+                RuleStatus ruleStatus = getRuleStatus(rUID);
+                if (ruleStatus == RuleStatus.NOT_INITIALIZED) {
+                    notInitailizedRules.add(rUID);
+                }
+
+            }
+        }
+        if (notInitailizedRules != null) {
+            for (String rUID : notInitailizedRules) {
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: Why not combine this for cycle with the above one?

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 5]
```diff
@@ -167,23 +159,96 @@ public RuleEngine() {
...
+    @Override
+    public void updated(ModuleType oldElement, ModuleType moduleType) {
+        // updating module types does not effect the rule
+    }
+
+    /**
+     * This method cleans the used by rule engine resources when the component deactivation is requested.
+     */
+    protected void dispose() {
+        for (Iterator< RuntimeRule>it = rules.values().iterator(); it.hasNext();) {
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: Missing: isDisposed = true;

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 6]
```diff
@@ -219,201 +288,432 @@ protected void updatedModuleHandlerFactory(ModuleHandlerFactory moduleHandlerFac
...
+    protected void runRule(RuntimeRule rule, RuleEngineCallbackImpl.TriggerData td) {
+        String rUID = rule.getUID();
+        RuleStatusInfo running = new RuleStatusInfo(RuleStatus.RUNNING);
+        synchronized (this) {
+            RuleStatus ruleStatus = getRuleStatus(rUID);
+            if (ruleStatus == RuleStatus.DISABLED || ruleStatus == RuleStatus.NOT_INITIALIZED) {
+                logger.error("Trying to execute rule ‘{}' with status '{}'", rUID, ruleStatus.getValue());
+                return;
+            }
+            if (ruleStatus == RuleStatus.RUNNING) {
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: Too complex logic. The rule has to be started only when it is in IDLE state and it must go back from RUNNING to IDLE when it finished its work. If it is disabled in the middle of rule execution (when it is in RUNNING state) It has to stop working and go into DISABLED state.

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 7]
```diff
@@ -219,201 +288,432 @@ protected void updatedModuleHandlerFactory(ModuleHandlerFactory moduleHandlerFac
...
+            if (isSatisfied) {
+                executeActions(rule);
+                logger.debug("The rule '{}' is executed.", rUID);
+            } else {
+                logger.debug("The rule '{}' is NOT executed, since it has unsatisfied conditions.", rUID);
+            }
+        } catch (Throwable t) {
+            logger.error("Fail to execute rule '{}': {}", new Object[] { rUID, t.getMessage() }, t);
+        }
+        RuleEngineCallbackImpl.TriggerData nextTrigger = null;
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: Too complex state changing.

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 8]
```diff
@@ -1000,7 +1035,7 @@ private void executeActions(Rule rule) {
         if (actions == null || actions.size() == 0) {
             return;
         }
-        for (Iterator<Action> it = actions.iterator(); it.hasNext();) {
+        for (Iterator< Action>it = actions.iterator(); it.hasNext();) {
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: same as above incorrect space

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java [hunk 9]
```diff
@@ -333,64 +257,37 @@ public Rule add(Rule rule) {
...
-            super.onAddElement(rule);
-        } catch (Exception e) {
-            logger.error("Can't add rule: {}", rule.getUID(), e);
+    protected void onAddElement(Rule rule) throws IllegalArgumentException {
+        Rule ruleWithUID = rule;
+        if (rule.getUID() == null) {
+            ruleWithUID = ruleEngine.initRuleId(rule);
+        }
+        String rUID = ruleWithUID.getUID();
+        postRuleAddedEvent(ruleWithUID);
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java]: The rule added event must be fire after adding the rule into ruleengine

File: bundles/automation/org.eclipse.smarthome.automation.integration.test/src/test/groovy/org/eclipse/smarthome/automation/integration/test/ProviderModuleTypes.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+import org.eclipse.smarthome.core.common.registry.Provider;
+import org.eclipse.smarthome.core.common.registry.ProviderChangeListener;
+
+/**
+ * This class is a {@link ModuleTypeProvider} test implementation.
+ *
+ * @author Ana Dimova
+ *
+ */
+public class ProviderModuleTypes implements Provider<ModuleType>, ModuleTypeProvider {
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.integration.test/src/test/groovy/org/eclipse/smarthome/automation/integration/test/ProviderModuleTypes.java]: May be better name is ModuleTypeProviderTestImpl or TestModuleTypeProviderImpl or just TestModuleTypeProvider, but there is a confusion between ProviderModuleTypes  and ModuleTypesProvider names

File: bundles/automation/org.eclipse.smarthome.automation.integration.test/src/test/groovy/org/eclipse/smarthome/automation/integration/test/ProviderModuleTypes.java [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+    private ModuleType[] moduleTypes;
+
+    public ProviderModuleTypes(ModuleType[] moduleTypes) {
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.integration.test/src/test/groovy/org/eclipse/smarthome/automation/integration/test/ProviderModuleTypes.java]: Same as confusion name as ProviderModuleType

File: bundles/automation/org.eclipse.smarthome.automation.integration.test/src/test/groovy/org/eclipse/smarthome/automation/integration/test/ProviderTemplates.java [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+import org.eclipse.smarthome.core.common.registry.Provider;
+import org.eclipse.smarthome.core.common.registry.ProviderChangeListener;
+
+/**
+ * This class is a {@link TemplateProvider} test implementation.
+ *
+ * @author Ana Dimova
+ *
+ */
+public class ProviderTemplates implements Provider<RuleTemplate>, TemplateProvider {
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.integration.test/src/test/groovy/org/eclipse/smarthome/automation/integration/test/ProviderTemplates.java]: same as above. May be TestTemplateProvider is better name

File: bundles/automation/org.eclipse.smarthome.automation.integration.test/src/test/groovy/org/eclipse/smarthome/automation/integration/test/ProviderModuleTypes.java (hunk 11 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.integration.test/src/test/groovy/org/eclipse/smarthome/automation/integration/test/ProviderModuleTypes.java]: This class will be removed after changing ModuleTypeProvider interface because then it will become unnecessary

File: bundles/automation/org.eclipse.smarthome.automation.integration.test/src/test/groovy/org/eclipse/smarthome/automation/integration/test/ProviderTemplates.java (hunk 12 above)
Comment (adimova) [inline bundles/automation/org.eclipse.smarthome.automation.integration.test/src/test/groovy/org/eclipse/smarthome/automation/integration/test/ProviderTemplates.java]: This class will be removed after changing TemplateProvider interface because then it will become unnecessary

Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.integration.test/src/test/groovy/org/eclipse/smarthome/automation/integration/test/ProviderTemplates.java]: Same as above, for now I think it is better to rename the test provider class

**Comment by danchom** [comment]
LGTM

**Comment by adimova** [comment]
There is a bug in this implementation of the RuleRegistry, I will fix it

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java [hunk 13]
```diff
@@ -333,24 +276,21 @@ public Rule add(Rule rule) {
...
                 }
-            } else {
-                ruleEngine.addRule(rule, false);
+                ruleUIDs.add(uid);
             }
-            super.onAddElement(rule);
-        } catch (Exception e) {
-            logger.error("Can't add rule: {}", rule.getUID(), e);
         }
+        postRuleAddedEvent(rule);
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java]: Move postRuleAddedEvent(rule) into added(Provider<Rule> provider, Rule element) method after call super.added()

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java [hunk 14]
```diff
@@ -412,46 +344,93 @@ protected void onRemoveElement(Rule rule) {
...
+        if (templateUID != null) {
+            synchronized (this) {
+                Set<String> ruleUIDs = mapTemplateToRules.get(templateUID);
+                if (ruleUIDs != null) {
+                    ruleUIDs.remove(uid);
+                }
+            }
         }
-        ruleEngine.updateRule(element);
+        postRuleUpdatedEvent(element, oldElement);
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java]: Same as above move it into updated

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java [hunk 15]
```diff
@@ -372,25 +312,17 @@ protected void postRuleStatusInfoEvent(RuleStatusInfo statusInfo, String ruleUID
...
+            synchronized (this) {
+                Set<String> ruleUIDs = mapTemplateToRules.get(templateUID);
+                if (ruleUIDs != null) {
+                    ruleUIDs.remove(uid);
                 }
             }
         }
-
-        super.onRemoveElement(rule);
+        postRuleRemovedEvent(rule);
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java]: Move it into removed method.

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java [hunk 16]
```diff
@@ -475,206 +453,116 @@ public Boolean isEnabled(String ruleUID) {
...
-    protected void validateConfiguration(Rule r) {
-        List<ConfigDescriptionParameter> configDescriptions = r.getConfigurationDescriptions();
-        Configuration moduleConfiguration = r.getConfiguration();
-        Map<String, Object> configuration = moduleConfiguration.getProperties();
-        if (configuration != null) {
-            validateConfiguration(configDescriptions, new HashMap<String, Object>(configuration));
-            handleModuleConfigReferences(r.getTriggers(), configuration);
-            handleModuleConfigReferences(r.getConditions(), configuration);
-            handleModuleConfigReferences(r.getActions(), configuration);
+    private void update(Provider<Rule> provider, Rule rule) {
```
Comment (danchom) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java]: Rename this method to updateRuleByTemplate, because update is not obvious what it is doing.

**Comment by danchom** [comment]
LGTM

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java [hunk 17]
```diff
@@ -1462,4 +1387,119 @@ public String getOutputName() {
...
+                Iterator<ConfigDescriptionParameter> i = configDescriptions.iterator();
+                while (i.hasNext()) {
+                    ConfigDescriptionParameter configParameter = i.next();
+                    if (i.hasNext()) {
+                        msg = msg + configParameter.getName() + "\", ";
+                    } else {
+                        msg = msg + configParameter.getName();
+                    }
+                }
+                logger.error("Extra configuration properties : " + msg + "\"!", new IllegalArgumentException());
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngine.java]: please use parameterized logging

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngineCallbackImpl.java [hunk 18]
```diff
@@ -42,22 +42,22 @@ protected RuleEngineCallbackImpl(RuleEngine re, RuntimeRule r) {
...
-                if (executor != null) {
-                    feature = executor.submit(new TriggerData(trigger, outputs));
-                }
+        synchronized (this) {
+            if (executor == null) {
+                return;
             }
+            future = executor.submit(new TriggerData(trigger, outputs));
         }
+        re.logger.debug("The trigger '{}' of rule '{}' is triggred.", trigger.getId(), r.getUID());
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleEngineCallbackImpl.java]: typo: triggred

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/type/ModuleTypeRegistryImpl.java [hunk 19]
```diff
@@ -83,10 +83,12 @@ public ModuleType get(String typeUID) {
...
         Collection<ModuleType> moduleTypes = null;
-        for (ModuleTypeProvider provider : providers) {
-            moduleTypes = provider.getModuleTypes(locale);
+        for (Provider<ModuleType> provider : elementMap.keySet()) {
+            if (provider instanceof ModuleTypeProvider) {
+                moduleTypes = ((ModuleTypeProvider) provider).getModuleTypes(locale);
+            }
             if (moduleTypes != null) {
-                for (Iterator<ModuleType> it = moduleTypes.iterator(); it.hasNext();) {
+                for (Iterator< ModuleType>it = moduleTypes.iterator(); it.hasNext();) {
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/type/ModuleTypeRegistryImpl.java]: revert the change of this line

File: bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/type/ModuleTypeRegistryImpl.java [hunk 20]
```diff
@@ -113,10 +115,12 @@ public ModuleType get(String typeUID) {
...
```
Comment (kaikreuzer) [inline bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/type/ModuleTypeRegistryImpl.java]: revert the change of this line

**Comment by danchom** [comment]
Ok thanks Kai, I'll merge it.

**Comment by maggu2810** [comment]
@adimova, @danchom Can you have a look at #2616 as it has been caused by this PR.

**Comment by maggu2810** [comment]
@danchom Please you squash and merge next time (see http://dev.eclipse.org/mhonarc/lists/smarthome-dev/msg00072.html and following)
